terraformignore.go

```diff
@@ -0,0 +1,219 @@
+package slug
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"text/scanner"
+)
+
+func parseIgnoreFile(rootPath string) []rule {
+	// Look for .terraformignore at our root path/src
+	file, err := os.Open(filepath.Join(rootPath, ".terraformignore"))
+	defer file.Close()
+
+	// If there's any kind of file error, punt and use the default ignore patterns
+	if err != nil {
+		// Only show the error debug if an error *other* than IsNotExist
+		if !os.IsNotExist(err) {
+			fmt.Printf("Error reading .terraformignore, default exclusions will apply: %v \n", err)
+		}
+		return defaultExclusions
+	}
+	return readRules(file)
+}
+
+func readRules(input io.Reader) []rule {
+	rules := defaultExclusions
+	scanner := bufio.NewScanner(input)
+	scanner.Split(bufio.ScanLines)
+
+	for scanner.Scan() {
+		pattern := scanner.Text()
+		// Ignore blank lines
+		if len(pattern) == 0 {
+			continue
+		}
+		// Trim spaces
+		pattern = strings.TrimSpace(pattern)
+		// Ignore comments
+		if pattern[0] == '#' {
+			continue
+		}
+		// New rule structure
+		rule := rule{}
+		// Exclusions
+		if pattern[0] == '!' {
+			rule.excluded = true
+			pattern = pattern[1:]
+		}
+		// If it is a directory, add ** so we catch descendants
+		if pattern[len(pattern)-1] == os.PathSeparator {
+			pattern = pattern + "**"
+		}
+		// If it starts with /, it is absolute
+		if pattern[0] == os.PathSeparator {
+			pattern = pattern[1:]
+		} else {
+			// Otherwise prepend **/
+			pattern = "**" + string(os.PathSeparator) + pattern
+		}
+		rule.val = pattern
+		rule.dirs = strings.Split(pattern, string(os.PathSeparator))
+		rules = append(rules, rule)
+	}
+
+	if err := scanner.Err(); err != nil {
+		fmt.Printf("Error reading .terraformignore, default exclusions will apply: %v \n", err)
+		return defaultExclusions
+	}
+	return rules
+}
+
+func matchIgnoreRule(path string, rules []rule) bool {
+	matched := false
+	path = filepath.FromSlash(path)
+	for _, rule := range rules {
+		match, _ := rule.match(path)
+
+		if match {
+			matched = !rule.excluded
+		}
+	}
+
+	if matched {
+		fmt.Printf("Skipping excluded path: %s \n", path)
```

Comment:
Should this also use the `debug` call?

Reply:
It will eventually be behind a flag, yes, while people are testing I have it noisy so people can see what's happening.